pkg/vm/cli/cli.go

```diff
@@ -1,197 +1,316 @@
+       boolean values, everything that can be converted to integer is treated as
+       integer and everything else is treated like a string.
 
-var commands = map[string]command{
-	"help":    {0, "show available commands", false},
-	"exit":    {0, "exit the VM prompt", false},
-	"ip":      {0, "show the current instruction", true},
-	"break":   {1, "place a breakpoint (> break 1)", true},
-	"estack":  {0, "show evaluation stack details", false},
-	"astack":  {0, "show alt stack details", false},
-	"istack":  {0, "show invocation stack details", false},
-	"loadavm": {1, "load an avm script into the VM (> load /path/to/script.avm)", false},
-	"loadhex": {1, "load a hex string into the VM (> loadhex 006166 )", false},
-	"loadgo":  {1, "compile and load a .go file into the VM (> load /path/to/file.go)", false},
-	"run":     {0, "execute the current loaded script", true},
-	"cont":    {0, "continue execution of the current loaded script", true},
-	"step":    {0, "step (n) instruction in the program (> step 10)", true},
-	"ops":     {0, "show the opcodes of the current loaded program", true},
+Passing parameters without operation is not supported. Parameters are packed
+into array before they're passed to the script, so effectively 'run' only
+supports contracts with signatures like this:
+   func Main(operation string, args []interface{}) interface{}
+
+Example:
+> run put string:"Something to put"`,
+		Func: handleRun,
+	},
+	{
+		Name: "cont",
+		Help: "Continue execution of the current loaded script",
+		LongHelp: "Continue execution of the current loaded script",
+		Func: handleCont,
+	},
+	{
+		Name: "step",
+		Help: "Step (n) instruction in the program",
+		LongHelp: `Usage: step [<n>]
+<n> is optional parameter to specify number of instructions to run, example:
+> step 10`,
+		Func: handleStep,
+	},
+	{
+		Name: "ops",
+		Help: "Dump opcodes of the current loaded program",
+		LongHelp: "Dump opcodes of the current loaded program",
+		Func: handleOps,
+	},
 }
 
-// VMCLI object for interacting with the VM.
+ // VMCLI object for interacting with the VM.
 type VMCLI struct {
-	vm *vm.VM
+	vm    *vm.VM
+	shell *ishell.Shell
 }
 
 // New returns a new VMCLI object.
 func New() *VMCLI {
-	return &VMCLI{
+	vmcli := VMCLI{
 		vm: vm.New(0),
+		shell: ishell.New(),
+	}
+	vmcli.shell.Set(VMKey, vmcli.vm)
+	for _, c := range commands {
+		vmcli.shell.AddCmd(c)
+	}
+	changePrompt(vmcli.shell, vmcli.vm)
+	return &vmcli
+}
+
+func getVMFromContext(c *ishell.Context) *vm.VM {
+	return c.Get(VMKey).(*vm.VM)
+}
+
+func checkVMIsReady(c *ishell.Context) bool {
+	v := getVMFromContext(c)
+	if v == nil || !v.Ready() {
+		c.Err(errors.New("VM is not ready: no program loaded"))
+		return false
 	}
+	return true
 }
 
-func (c *VMCLI) handleCommand(cmd string, args ...string) {
-	com, ok := commands[cmd]
-	if !ok {
-		fmt.Printf("unknown command (%s)\n", cmd)
+func handleExit(c *ishell.Context) {
```

Comment:
Looked through and there is a lot 'handle%' functions. What do you think about moving them to some Handler responsible object in order to make life easier at some point when we will cover this one with unit tests?

Reply:
Not sure what do you mean by that. These are handlers and they have to satisfy some specific signature that `ishell` expects.

Reply:
I thought about moving all handle functions to separate file and have type 'Handler' which is responsible for handling

Reply:
probably after that cli will be very thin so ... maybe it's not the case for moving them